### DISCUSSION TITLE
building with configuration VC-WIN64A fails

### DISCUSSION THREAD

**Discussion by jamuir** [openssl/openssl#22385] [Q&A] [answered]
I tried to build `master` (eaee1765) using the configuration `VC-WIN64A` on two different x86_64 windows machines and both builds failed.

I used a bat file containing the following lines:

```
REM doit.bat > log.txt 2>&1
perl Configure VC-WIN64A --prefix="C:\Users\tvavra\source\repos\james\tmp\beforespace afterspace"
nmake
nmake test
nmake install
```

The first `nmake` step fails with the following error:

```
	cmd /C ""link" /nologo /debug /dll  /nologo /debug @C:\Users\tvavra\AppData\Local\Temp\2\nmB0A4.tmp /implib:libcrypto.lib || (DEL /Q libcrypto-3-x64.* libcrypto.lib & EXIT 1)"
crypto\aes\libcrypto-shlib-aes_cfb.obj : fatal error LNK1112: module machine type 'x86' conflicts with target machine type 'x64'
Could Not Find C:\Users\tvavra\source\repos\james\openssl\libcrypto-3-x64.*
NMAKE : fatal error U1077: 'cmd /C ""link" /nologo /debug /dll  /nologo /debug @C:\Users\tvavra\AppData\Local\Temp\2\nmB0A4.tmp /implib:libcrypto.lib || (DEL /Q libcrypto-3-x64.* libcrypto.lib & EXIT 1)"' : return code '0x1'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX86\x86\nmake.exe" /                   _tests' : return code '0x2'
Stop.
```

Then both `nmake test` and `nmake install` fail with the same error.

**Comment by jamuir** [comment]
I should note that if I call `Configure` without `VC-WIN64A` (so that the script selects its own build-config), then the build completes successfully.

**Comment by mattcaswell** [comment, marked as answer]
> fatal error LNK1112: module machine type 'x86' conflicts with target machine type 'x64'

This normally occurs because of a discrepancy between the toolchain being used and the architecture that you are attempting to build for, i.e. using the x86 VC toolchain but attempting to build for VC-WIN64A (i.e. x86_64).

If you run the compiler (`cl.exe`) by itself with no arguments it should tell you what architecture it is targeting.

See:

https://github.com/openssl/openssl/blob/eaee1765a49c6a8ba728e3e2d18bb67bff8aaa55/NOTES-WINDOWS.md?plain=1#L65-L73

**Comment by jamuir** [comment]
thanks for your reply, Matt.

you were right that I did not have the VC toolchain initialized correctly.

I set it correctly now like so:

```
"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build"\vcvarsall.bat amd64
```
`cl` reports it is targeting `x64`.

Unfortunately, I am still not able to build using VC-WIN64A.

```
@REM doit.bat > log.txt 2>&1

set path=C:\Program Files\NASM;%path%
nasm -v
perl -v

cl

perl Configure VC-WIN64A --prefix="C:\Users\muirj\source\repos\tmp\beforespace afterspace"
nmake
```

I hit the following failure in the `lib` tool:

```
        "lib" /nologo /out:apps\libapps.lib @C:\Users\muirj\AppData\Local\Temp\nmEFDA.tmp
apps\lib\libapps-lib-fmt.obj : fatal error LNK1112: module machine type 'x86' conflicts with target machine type 'x64'
NMAKE : fatal error U1077: '"lib" /nologo /out:apps\libapps.lib @C:\Users\muirj\AppData\Local\Temp\nmEFDA.tmp' : return code '0x458'       
Stop.
NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64\nmake.exe" /
                   _build_sw' : return code '0x2'
```

it seems like the option `/machine` is missing from `lib`.

**Reply by mattcaswell** [reply]
Did you do a "make clean" between builds? I suspect some leftovers from your previous failed build. Maybe even try a fresh clone.

**Reply by jamuir** [reply]
thanks again, Matt.

You were right.  It seems to have been some artifact from my previous failed build causing the problem.  I thought I had done `nmake clean` after the failure, but it seems likely I did not.

After a `nmake clean` and a few calls to `git clean -x ...` I went back to a completely clean tree, and then I was finally able to build successfully using `VC-WIN64A`.

**Comment by laultman** [comment]
This helped me too. I followed the [VS2022 instructions to set the Developer VC environment](https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170) rather than trying to manually set it. Also, rather than trying to clean up after a failed build, I used @jamuir hint and simply deleted the old one and unzipped a fresh one.

**Comment by laultman** [comment]
@mattcaswell Are the generated header files the same between builds or are they different? I am building the [Microsoft.cpprestskd](https://github.com/microsoft/cpprestsdk) adding ARM64 UWP support. This has forced me to also build the OpenSSL project. I'm only supporting only x64 and ARM64 and ARM64-UWP. These are for industrial and military applications.

**Reply by mattcaswell** [reply]
They may be different between different builds

**Comment by Nick7K** [comment]
Note that there is also a ticket open for issues with the OpenSSL under vcpkg arm64-windows

[https://github.com/microsoft/vcpkg/issues/42828](url)

**Comment by sergeevabc** [comment]
What's the difference between VC-WIN64A, VC-WIN64A-rtt and VC-WIN64A-HYBRIDCRT?

**Reply by levitte** [reply]
VC-WIN64A-rtt isn't a configuration we know.

VC-WIN64A-HYBRIDCRT is a config to link with a different C runtime library, shown by these lines:

https://github.com/openssl/openssl/blob/49a3e7adc392a45c3fba93f5759df6797fcd8238/Configurations/50-win-hybridcrt.conf#L25-L26

**Reply by sergeevabc** [reply]
Re: @levitte

Richard, I use and will continue to use Windows 7 x64. Fortunately, OpenSSL is compiled in such a way that it runs on this OS without issues. This is great news for many souls who don't live on the Microsoft's EOL schedule. Kudos to you, the developers of OpenSSL, who stick to the mission.

![](https://github.com/user-attachments/assets/33351ab9-01db-492a-b62d-0712736f843c)

However, there is one thing that confuses me — too many valid distributions.

* [Firedaemon](https://kb.firedaemon.com/support/solutions/articles/4000121705-openssl-binary-distributions-for-microsoft-windows#Download-OpenSSL)
* [Shining Light](https://slproweb.com/products/Win32OpenSSL.html)
* [Internet Component Suite](https://wiki.overbyte.eu/wiki/index.php/ICS_Download#Download_OpenSSL_Binaries)

Which one should I use?

**Reply by levitte** [reply]
Of those three, I only knew Shining Light. I can't give much advice in that regard, other than to try them out and see what sticks.

**Reply by trueqbit** [reply]
> However, there is one thing that confuses me — too many valid distributions.

These are binaries from individuals/companies who want to contribute to the community with the binaries they build and consume themselves.
You'd have to read the details about what to expect (or just try them out).

I can only comment on the FireDaemon distribution, which is basically a "developer" build for integration into your own product:
* It doesn't automatically load the global OpenSSL configuration, allowing you to set up your application in your own way,
* It is built with the VC_WIN64A-HYBRIDCRT triple, hence has no dependency on the Visual C Runtime (redistributable), only on the "Universal" CRT, which makes distributing your application much simpler.

There is a decision process within the OpenSSL team as to which "blessed" Windows distributions should be offered in the future. The result of a [recent survey](https://openssl-communities.org/d/hLGryc76/openssl-library-binary-distributions-and-installers-for-microsoft-windows-poll/) might give an outlook on what might be part of it..
